test/e2e/etcd_multi_node_test.go

```diff
@@ -0,0 +1,441 @@
+			By("Scale up of a healthy cluster (from 1->3)")
+			ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+			etcd.Spec.Replicas = multiNodeEtcdReplicas
+			updateAndCheckEtcd(ctx, cl, objLogger, etcd)
+
+			// TODO: Uncomment me once scale down replicas from 3 to 1 is supported.
+			// By("Scale down of a healthy cluster (from 3 to 1)")
+			// ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+			// etcd.Spec.Replicas = replicas
+			// updateAndCheckEtcd(ctx, cl, objLogger, etcd)
+
+			By("Delete single-node etcd")
+			deleteAndCheckEtcd(ctx, cl, objLogger, etcd)
+		})
+	})
+})
+
+// checkEventuallyEtcdRollingUpdateDone is a helper function, uses Gomega Eventually.
+// Returns the function until etcd rolling updates is done for given timeout and polling interval or
+// it raises assertion error.
+//
+// checkEventuallyEtcdRollingUpdateDone ensures rolling updates of etcd resources(etcd/sts) is done.
+func checkEventuallyEtcdRollingUpdateDone(ctx context.Context, cl client.Client, logger logr.Logger, etcd *v1alpha1.Etcd,
+	oldStsObservedGeneration, oldEtcdObservedGeneration int64) {
+	Eventually(func() error {
+		sts := &appsv1.StatefulSet{}
+		ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), sts)).To(Succeed())
+
+		if sts.Status.ObservedGeneration <= oldStsObservedGeneration {
+			return fmt.Errorf("waiting for statefulset rolling update to complete %d pods at revision %s",
+				sts.Status.UpdatedReplicas, sts.Status.UpdateRevision)
+		}
+
+		if sts.Status.UpdatedReplicas != *sts.Spec.Replicas {
+			return fmt.Errorf("waiting for statefulset rolling update to complete, UpdatedReplicas is %d, but expected to be %d",
+				sts.Status.UpdatedReplicas, *sts.Spec.Replicas)
+		}
+
+		ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+		if *etcd.Status.ObservedGeneration <= oldEtcdObservedGeneration {
+			return fmt.Errorf("waiting for etcd %q rolling update to complete", etcd.Name)
+		}
+		return nil
+	}, timeout*3, pollingInterval).Should(BeNil())
+}
+
+// hibernateAndCheckEtcd scales down etcd replicas to zero and ensures
+func hibernateAndCheckEtcd(ctx context.Context, cl client.Client, logger logr.Logger, etcd *v1alpha1.Etcd) {
+	ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), etcd)).To(Succeed())
+	etcd.Spec.Replicas = 0
+	etcd.SetAnnotations(
+		map[string]string{
+			v1beta1constants.GardenerOperation: v1beta1constants.GardenerOperationReconcile,
+		})
+	ExpectWithOffset(1, cl.Update(ctx, etcd)).ShouldNot(HaveOccurred())
+	logger.Info("Waiting to hibernate")
+
+	logger.Info("Checking statefulset")
+	Eventually(func() error {
+		sts := &appsv1.StatefulSet{}
+		ExpectWithOffset(1, cl.Get(ctx, client.ObjectKeyFromObject(etcd), sts)).To(Succeed())
+
+		if sts.Status.ReadyReplicas != 0 {
+			return fmt.Errorf("sts %s not ready", etcd.Name)
+		}
+		return nil
+	}, timeout*3, pollingInterval).Should(BeNil())
```

Comment:
Can you check the `statefulset` after checking the `etcd` resource? This can then also be done without `Eventually` because the `etcd` resource should only meet the requirements after the `statefulset` is ready.